File: src/double_ok_gesture/capture_gate.py
"""Pre-capture gate for glasses-side double OK collection."""

from __future__ import annotations

import json
import math
import shutil
import subprocess
import time
from dataclasses import dataclass, replace
from enum import Enum
from pathlib import Path
from typing import Any

import numpy as np

from .recognizer import DoubleOKResult


class GateReason(str, Enum):
    READY = "ready"
    GLASSES_POSE_MISSING = "glasses_pose_missing"
    GLASSES_POSE_BAD = "glasses_pose_bad"
    NEED_TWO_HANDS = "need_two_hands"
    HANDS_OUT_OF_FRAME = "hands_out_of_frame"
    HANDS_NOT_CENTERED = "hands_not_centered"
    HANDS_TOO_CLOSE = "hands_too_close"
    NEED_DOUBLE_OK = "need_double_ok"
    AVOID_DOUBLE_OK = "avoid_double_ok"


class StereoGateMode(str, Enum):
    LEFT = "left"
    BOTH = "both"


@dataclass(frozen=True)
class GlassesPose:
    """Glasses pose angles in degrees from the glasses IMU or device pose stack."""

    pitch: float | None = None
    roll: float | None = None
    yaw: float | None = None


@dataclass(frozen=True)
class CaptureGateConfig:
    """Tunable capture-start constraints in normalized camera coordinates."""

    require_glasses_pose: bool = False
    pitch_min: float = -20.0
    pitch_max: float = 20.0
    roll_min: float = -12.0
    roll_max: float = 12.0
    yaw_min: float = -25.0
    yaw_max: float = 25.0
    frame_margin: float = 0.03
    center_x_min: float = 0.20
    center_x_max: float = 0.80
    center_y_min: float = 0.18
    center_y_max: float = 0.82
    min_hand_separation: float = 0.16
    use_stable_double_ok: bool = True
    require_double_ok: bool = True

    def __post_init__(self) -> None:
        _validate_range("pitch", self.pitch_min, self.pitch_max)
        _validate_range("roll", self.roll_min, self.roll_max)
        _validate_range("yaw", self.yaw_min, self.yaw_max)
        _validate_normalized_range("center_x", self.center_x_min, self.center_x_max)
        _validate_normalized_range("center_y", self.center_y_min, self.center_y_max)
        if not math.isfinite(self.frame_margin) or not 0.0 <= self.frame_margin < 0.5:
            raise ValueError("frame_margin must be finite and in [0.0, 0.5)")
        if not math.isfinite(self.min_hand_separation) or self.min_hand_separation < 0.0:
            raise ValueError("min_hand_separation must be finite and non-negative")
        if not isinstance(self.use_stable_double_ok, bool):
            raise ValueError("use_stable_double_ok must be a boolean")
        if not isinstance(self.require_double_ok, bool):
            raise ValueError("require_double_ok must be a boolean")


@dataclass(frozen=True)
class CaptureGateDecision:
    ready: bool
    reason: GateReason
    prompt: str
    glasses_pose_ok: bool
    hands_centered: bool
    hands_visible: bool
    hands_separated: bool
    double_ok: bool
    gesture_ok: bool
    hand_count: int


@dataclass(frozen=True)
class StereoCaptureGateDecision:
    ready: bool
    reason: GateReason
    prompt: str
    left: CaptureGateDecision
    right: CaptureGateDecision | None = None


PROMPTS = {
    GateReason.READY: "条件满足，开始采集",
    GateReason.GLASSES_POSE_MISSING: "等待眼镜姿态数据",
    GateReason.GLASSES_POSE_BAD: "请调整眼镜角度，保持视野正对双手",
    GateReason.NEED_TWO_HANDS: "请把双手放入相机画面",
    GateReason.HANDS_OUT_OF_FRAME: "请把双手完整放入画面",
    GateReason.HANDS_NOT_CENTERED: "请把双手移到画面中心",
    GateReason.HANDS_TOO_CLOSE: "请将双手分开一些",
    GateReason.NEED_DOUBLE_OK: "请双手分开并做出 OK 手势",
    GateReason.AVOID_DOUBLE_OK: "负样本采集中，请不要同时做双手 OK",
}

OVERLAY_LABELS = {
    GateReason.READY: "READY_TO_CAPTURE",
    GateReason.GLASSES_POSE_MISSING: "WAITING_GLASSES_POSE",
    GateReason.GLASSES_POSE_BAD: "ADJUST_GLASSES_ANGLE",
    GateReason.NEED_TWO_HANDS: "SHOW_TWO_HANDS",
    GateReason.HANDS_OUT_OF_FRAME: "HANDS_OUT_OF_FRAME",
    GateReason.HANDS_NOT_CENTERED: "MOVE_HANDS_TO_CENTER",
    GateReason.HANDS_TOO_CLOSE: "SEPARATE_HANDS",
    GateReason.NEED_DOUBLE_OK: "MAKE_DOUBLE_OK",
    GateReason.AVOID_DOUBLE_OK: "AVOID_DOUBLE_OK",
}


def load_glasses_pose(path: str | Path | None) -> GlassesPose | None:
    if not path:
        return None
    pose_path = Path(path)
    if not pose_path.exists():
        return None
    try:
        with pose_path.open("r", encoding="utf-8") as file:
            data: Any = json.load(file)
        if not isinstance(data, dict):
            return None
        return GlassesPose(
            pitch=_optional_float(data.get("pitch")),
            roll=_optional_float(data.get("roll")),
            yaw=_optional_float(data.get("yaw")),
        )
    except (OSError, json.JSONDecodeError, TypeError, ValueError):
        # The glasses process may be replacing this file while a frame is read.
        return None


def evaluate_capture_gate(
    result: DoubleOKResult,
    config: CaptureGateConfig | None = None,
    glasses_pose: GlassesPose | None = None,
) -> CaptureGateDecision:
    cfg = config or CaptureGateConfig()
    pose_ok, pose_reason = _check_glasses_pose(glasses_pose, cfg)
    hand_count = len(result.hands)
    hands_visible = False
    hands_centered = False
    hands_separated = False
    double_ok = result.double_ok and (result.stable_double_ok if cfg.use_stable_double_ok else True)

    if not pose_ok:
        return _decision(
            pose_reason,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    if hand_count < 2:
        return _decision(
            GateReason.NEED_TWO_HANDS,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    landmarks = [_landmark_xy(hand.landmarks) for hand in result.hands[:2]]
    if any(points is None for points in landmarks):
        return _decision(
            GateReason.HANDS_OUT_OF_FRAME,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )
    landmarks = [points for points in landmarks if points is not None]
    hands_visible = all(_is_hand_visible(points, cfg) for points in landmarks)
    if not hands_visible:
        return _decision(
            GateReason.HANDS_OUT_OF_FRAME,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    centers = np.asarray([points.mean(axis=0) for points in landmarks], dtype=np.float32)
    hands_centered = all(_is_point_centered(center, cfg) for center in centers)
    if not hands_centered:
        return _decision(
            GateReason.HANDS_NOT_CENTERED,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    hands_separated = float(np.linalg.norm(centers[0] - centers[1])) >= cfg.min_hand_separation
    if not hands_separated:
        return _decision(
            GateReason.HANDS_TOO_CLOSE,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    if cfg.require_double_ok and not double_ok:
        return _decision(
            GateReason.NEED_DOUBLE_OK,
            cfg,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count,
        )

    return _decision(
        GateReason.READY,
        cfg,
        pose_ok,
        hands_centered,
        hands_visible,
        hands_separated,
        double_ok,
        hand_count,
    )


def evaluate_labeled_capture_gate(
    result: DoubleOKResult,
    label: str,
    config: CaptureGateConfig | None = None,
    glasses_pose: GlassesPose | None = None,
) -> CaptureGateDecision:
    """Evaluate geometry plus the gesture condition required by a sample label."""

    cfg = config or CaptureGateConfig()
    if label == "double_ok":
        return evaluate_capture_gate(result, cfg, glasses_pose)
    if label != "not_double_ok":
        raise ValueError(f"Unsupported capture label: {label}")

    geometry_cfg = replace(cfg, require_double_ok=False)
    geometry_decision = evaluate_capture_gate(result, geometry_cfg, glasses_pose)
    if not geometry_decision.ready:
        return replace(geometry_decision, gesture_ok=not result.double_ok)

    common = (
        geometry_cfg,
        geometry_decision.glasses_pose_ok,
        geometry_decision.hands_centered,
        geometry_decision.hands_visible,
        geometry_decision.hands_separated,
        result.double_ok,
        geometry_decision.hand_count,
    )
    if result.double_ok:
        return _decision(GateReason.AVOID_DOUBLE_OK, *common, gesture_ok=False)
    return _decision(GateReason.READY, *common, gesture_ok=True)


def evaluate_stereo_capture_gate(
    left_result: DoubleOKResult,
    right_result: DoubleOKResult | None = None,
    config: CaptureGateConfig | None = None,
    glasses_pose: GlassesPose | None = None,
    mode: StereoGateMode | str = StereoGateMode.LEFT,
) -> StereoCaptureGateDecision:
    """Combine capture gates for a binocular camera pair.

    LEFT mode keeps the early prototype path simple by using the left image as
    the gate authority. BOTH mode requires both views to satisfy the same gate.
    """

    stereo_mode = StereoGateMode(mode)
    left_decision = evaluate_capture_gate(left_result, config, glasses_pose)
    if stereo_mode == StereoGateMode.LEFT:
        return StereoCaptureGateDecision(
            ready=left_decision.ready,
            reason=left_decision.reason,
            prompt=left_decision.prompt,
            left=left_decision,
        )

    if right_result is None:
        raise ValueError("right_result is required when stereo gate mode is 'both'")

    right_decision = evaluate_capture_gate(right_result, config, glasses_pose)
    if left_decision.ready and right_decision.ready:
        return StereoCaptureGateDecision(
            ready=True,
            reason=GateReason.READY,
            prompt=PROMPTS[GateReason.READY],
            left=left_decision,
            right=right_decision,
        )

    if not left_decision.ready and not right_decision.ready:
        prompt = (
            f"双目: {left_decision.prompt}"
            if left_decision.reason == right_decision.reason
            else f"左眼: {left_decision.prompt}; 右眼: {right_decision.prompt}"
        )
        return StereoCaptureGateDecision(
            ready=False,
            reason=left_decision.reason,
            prompt=prompt,
            left=left_decision,
            right=right_decision,
        )

    blocked_view = "左眼" if not left_decision.ready else "右眼"
    blocked_decision = left_decision if not left_decision.ready else right_decision
    return StereoCaptureGateDecision(
        ready=False,
        reason=blocked_decision.reason,
        prompt=f"{blocked_view}: {blocked_decision.prompt}",
        left=left_decision,
        right=right_decision,
    )


class PromptSpeaker:
    """Throttle repeated prompts and optionally speak them through espeak."""

    def __init__(self, enabled: bool = False, min_interval_sec: float = 2.0) -> None:
        if not math.isfinite(min_interval_sec) or min_interval_sec < 0.0:
            raise ValueError("min_interval_sec must be finite and non-negative")
        self.enabled = enabled
        self.min_interval_sec = min_interval_sec
        self._last_prompt: str | None = None
        self._last_time = 0.0
        self._tts = shutil.which("espeak") if enabled else None

    def emit(self, prompt: str) -> None:
        now = time.monotonic()
        if prompt == self._last_prompt and now - self._last_time < self.min_interval_sec:
            return
        self._last_prompt = prompt
        self._last_time = now
        print(prompt)
        if self._tts:
            subprocess.Popen(
                [self._tts, "-v", "zh", prompt],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )


def draw_gate_overlay(frame_bgr: np.ndarray, decision: CaptureGateDecision, config: CaptureGateConfig) -> None:
    import cv2

    h, w = frame_bgr.shape[:2]
    x1 = int(config.center_x_min * w)
    x2 = int(config.center_x_max * w)
    y1 = int(config.center_y_min * h)
    y2 = int(config.center_y_max * h)
    color = (40, 200, 40) if decision.ready else (40, 160, 255)
    cv2.rectangle(frame_bgr, (x1, y1), (x2, y2), color, 2)
    cv2.putText(frame_bgr, OVERLAY_LABELS[decision.reason], (24, 82), cv2.FONT_HERSHEY_SIMPLEX, 0.75, color, 2)
    checks = [
        ("POSE", decision.glasses_pose_ok),
        ("VISIBLE", decision.hands_visible),
        ("CENTER", decision.hands_centered),
        ("SEPARATE", decision.hands_separated),
        ("GESTURE", decision.gesture_ok),
    ]
    for index, (name, passed) in enumerate(checks):
        check_color = (40, 200, 40) if passed else (40, 160, 255)
        marker = "OK" if passed else "--"
        cv2.putText(
            frame_bgr,
            f"{name}:{marker}",
            (24, 112 + index * 24),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.52,
            check_color,
            1,
        )


def _check_glasses_pose(
    glasses_pose: GlassesPose | None,
    config: CaptureGateConfig,
) -> tuple[bool, GateReason]:
    if not config.require_glasses_pose:
        return True, GateReason.READY
    if glasses_pose is None or any(
        value is None for value in (glasses_pose.pitch, glasses_pose.roll, glasses_pose.yaw)
    ):
        return False, GateReason.GLASSES_POSE_MISSING
    checks = [
        _in_optional_range(glasses_pose.pitch, config.pitch_min, config.pitch_max),
        _in_optional_range(glasses_pose.roll, config.roll_min, config.roll_max),
        _in_optional_range(glasses_pose.yaw, config.yaw_min, config.yaw_max),
    ]
    if all(checks):
        return True, GateReason.READY
    return False, GateReason.GLASSES_POSE_BAD


def _decision(
    reason: GateReason,
    _config: CaptureGateConfig,
    glasses_pose_ok: bool,
    hands_centered: bool,
    hands_visible: bool,
    hands_separated: bool,
    double_ok: bool,
    hand_count: int,
    *,
    gesture_ok: bool | None = None,
) -> CaptureGateDecision:
    if gesture_ok is None:
        gesture_ok = double_ok if _config.require_double_ok else True
    return CaptureGateDecision(
        ready=reason == GateReason.READY,
        reason=reason,
        prompt=PROMPTS[reason],
        glasses_pose_ok=glasses_pose_ok,
        hands_centered=hands_centered,
        hands_visible=hands_visible,
        hands_separated=hands_separated,
        double_ok=double_ok,
        gesture_ok=gesture_ok,
        hand_count=hand_count,
    )


def _is_hand_visible(points: np.ndarray, config: CaptureGateConfig) -> bool:
    margin = config.frame_margin
    return bool(
        np.all(points[:, 0] >= margin)
        and np.all(points[:, 0] <= 1.0 - margin)
        and np.all(points[:, 1] >= margin)
        and np.all(points[:, 1] <= 1.0 - margin)
    )


def _landmark_xy(landmarks: object) -> np.ndarray | None:
    try:
        points = np.asarray(landmarks, dtype=np.float32)
    except (TypeError, ValueError):
        return None
    if points.ndim != 2 or points.shape[0] != 21 or points.shape[1] < 2:
        return None
    points = points[:, :2]
    if not np.isfinite(points).all():
        return None
    return points


def _is_point_centered(point: np.ndarray, config: CaptureGateConfig) -> bool:
    x, y = float(point[0]), float(point[1])
    return config.center_x_min <= x <= config.center_x_max and config.center_y_min <= y <= config.center_y_max


def _in_optional_range(value: float | None, minimum: float, maximum: float) -> bool:
    if value is None:
        return True
    return minimum <= value <= maximum


def _optional_float(value: object) -> float | None:
    if value is None:
        return None
    result = float(value)
    if not math.isfinite(result):
        raise ValueError("pose values must be finite")
    return result


def _validate_range(name: str, minimum: float, maximum: float) -> None:
    if not math.isfinite(minimum) or not math.isfinite(maximum) or minimum > maximum:
        raise ValueError(f"{name}_min and {name}_max must be finite and ordered")


def _validate_normalized_range(name: str, minimum: float, maximum: float) -> None:
    _validate_range(name, minimum, maximum)
    if not 0.0 <= minimum < maximum <= 1.0:
        raise ValueError(f"{name}_min and {name}_max must satisfy 0 <= min < max <= 1")

File: src/capture_gate.cpp
#include "double_ok_gesture/capture_gate.hpp"

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <fstream>
#include <iostream>
#include <regex>
#include <stdexcept>

namespace double_ok_gesture {
namespace {

struct Point2 {
    double x = 0.0;
    double y = 0.0;
};

void validate_range(const std::string& name, double minimum, double maximum) {
    if (!std::isfinite(minimum) || !std::isfinite(maximum) || minimum > maximum) {
        throw std::invalid_argument(name + "_min and " + name + "_max must be finite and ordered");
    }
}

void validate_normalized_range(const std::string& name, double minimum, double maximum) {
    validate_range(name, minimum, maximum);
    if (!(0.0 <= minimum && minimum < maximum && maximum <= 1.0)) {
        throw std::invalid_argument(name + "_min and " + name + "_max must satisfy 0 <= min < max <= 1");
    }
}

std::optional<double> find_number(const std::string& text, const std::string& key) {
    const std::regex pattern("\"" + key + "\"\\s*:\\s*(-?(?:\\d+\\.?\\d*|\\.\\d+)(?:[eE][+-]?\\d+)?)");
    std::smatch match;
    if (!std::regex_search(text, match, pattern)) {
        return std::nullopt;
    }
    const double value = std::stod(match[1].str());
    if (!std::isfinite(value)) {
        throw std::runtime_error("pose values must be finite");
    }
    return value;
}

std::pair<bool, GateReason> check_glasses_pose(
    const std::optional<GlassesPose>& glasses_pose,
    const CaptureGateConfig& config) {
    if (!config.require_glasses_pose) {
        return {true, GateReason::Ready};
    }
    if (!glasses_pose || !glasses_pose->pitch || !glasses_pose->roll || !glasses_pose->yaw) {
        return {false, GateReason::GlassesPoseMissing};
    }
    const bool ok = *glasses_pose->pitch >= config.pitch_min && *glasses_pose->pitch <= config.pitch_max &&
                    *glasses_pose->roll >= config.roll_min && *glasses_pose->roll <= config.roll_max &&
                    *glasses_pose->yaw >= config.yaw_min && *glasses_pose->yaw <= config.yaw_max;
    return {ok, ok ? GateReason::Ready : GateReason::GlassesPoseBad};
}

bool is_hand_visible(const Landmarks& points, const CaptureGateConfig& config) {
    const double margin = config.frame_margin;
    return std::all_of(points.begin(), points.end(), [margin](const Point3& point) {
        return point.x >= margin && point.x <= 1.0 - margin && point.y >= margin && point.y <= 1.0 - margin;
    });
}

Point2 center_of(const Landmarks& points) {
    Point2 center;
    for (const Point3& point : points) {
        center.x += point.x;
        center.y += point.y;
    }
    center.x /= static_cast<double>(points.size());
    center.y /= static_cast<double>(points.size());
    return center;
}

bool is_point_centered(const Point2& point, const CaptureGateConfig& config) {
    return point.x >= config.center_x_min && point.x <= config.center_x_max && point.y >= config.center_y_min &&
           point.y <= config.center_y_max;
}

double distance(const Point2& lhs, const Point2& rhs) {
    const double dx = lhs.x - rhs.x;
    const double dy = lhs.y - rhs.y;
    return std::sqrt(dx * dx + dy * dy);
}

CaptureGateDecision decision(
    GateReason reason,
    const CaptureGateConfig& config,
    bool glasses_pose_ok,
    bool hands_centered,
    bool hands_visible,
    bool hands_separated,
    bool double_ok,
    int hand_count,
    std::optional<bool> gesture_ok = std::nullopt) {
    const bool final_gesture_ok = gesture_ok.value_or(config.require_double_ok ? double_ok : true);
    return {
        reason == GateReason::Ready,
        reason,
        gate_prompt(reason),
        glasses_pose_ok,
        hands_centered,
        hands_visible,
        hands_separated,
        double_ok,
        final_gesture_ok,
        hand_count,
    };
}

double monotonic_seconds() {
    using Clock = std::chrono::steady_clock;
    return std::chrono::duration<double>(Clock::now().time_since_epoch()).count();
}

}  // namespace

void CaptureGateConfig::validate() const {
    validate_range("pitch", pitch_min, pitch_max);
    validate_range("roll", roll_min, roll_max);
    validate_range("yaw", yaw_min, yaw_max);
    validate_normalized_range("center_x", center_x_min, center_x_max);
    validate_normalized_range("center_y", center_y_min, center_y_max);
    if (!std::isfinite(frame_margin) || frame_margin < 0.0 || frame_margin >= 0.5) {
        throw std::invalid_argument("frame_margin must be finite and in [0.0, 0.5)");
    }
    if (!std::isfinite(min_hand_separation) || min_hand_separation < 0.0) {
        throw std::invalid_argument("min_hand_separation must be finite and non-negative");
    }
}

const char* gate_reason_value(GateReason reason) {
    switch (reason) {
        case GateReason::Ready:
            return "ready";
        case GateReason::GlassesPoseMissing:
            return "glasses_pose_missing";
        case GateReason::GlassesPoseBad:
            return "glasses_pose_bad";
        case GateReason::NeedTwoHands:
            return "need_two_hands";
        case GateReason::HandsOutOfFrame:
            return "hands_out_of_frame";
        case GateReason::HandsNotCentered:
            return "hands_not_centered";
        case GateReason::HandsTooClose:
            return "hands_too_close";
        case GateReason::NeedDoubleOK:
            return "need_double_ok";
        case GateReason::AvoidDoubleOK:
            return "avoid_double_ok";
    }
    return "unknown";
}

const char* gate_prompt(GateReason reason) {
    switch (reason) {
        case GateReason::Ready:
            return "条件满足，开始采集";
        case GateReason::GlassesPoseMissing:
            return "等待眼镜姿态数据";
        case GateReason::GlassesPoseBad:
            return "请调整眼镜角度，保持视野正对双手";
        case GateReason::NeedTwoHands:
            return "请把双手放入相机画面";
        case GateReason::HandsOutOfFrame:
            return "请把双手完整放入画面";
        case GateReason::HandsNotCentered:
            return "请把双手移到画面中心";
        case GateReason::HandsTooClose:
            return "请将双手分开一些";
        case GateReason::NeedDoubleOK:
            return "请双手分开并做出 OK 手势";
        case GateReason::AvoidDoubleOK:
            return "负样本采集中，请不要同时做双手 OK";
    }
    return "未知状态";
}

GateReason gate_reason_from_string(const std::string& value) {
    const std::array<GateReason, 9> reasons = {
        GateReason::Ready,
        GateReason::GlassesPoseMissing,
        GateReason::GlassesPoseBad,
        GateReason::NeedTwoHands,
        GateReason::HandsOutOfFrame,
        GateReason::HandsNotCentered,
        GateReason::HandsTooClose,
        GateReason::NeedDoubleOK,
        GateReason::AvoidDoubleOK,
    };
    for (GateReason reason : reasons) {
        if (value == gate_reason_value(reason)) {
            return reason;
        }
    }
    throw std::invalid_argument("Unsupported gate reason: " + value);
}

StereoGateMode stereo_gate_mode_from_string(const std::string& value) {
    if (value == "left") {
        return StereoGateMode::Left;
    }
    if (value == "both") {
        return StereoGateMode::Both;
    }
    throw std::invalid_argument("Unsupported stereo gate mode: " + value);
}

std::optional<GlassesPose> load_glasses_pose(const std::filesystem::path& path) {
    if (path.empty() || !std::filesystem::exists(path)) {
        return std::nullopt;
    }
    try {
        std::ifstream in(path);
        if (!in) {
            return std::nullopt;
        }
        const std::string text((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
        return GlassesPose{
            find_number(text, "pitch"),
            find_number(text, "roll"),
            find_number(text, "yaw"),
        };
    } catch (const std::exception&) {
        return std::nullopt;
    }
}

CaptureGateDecision evaluate_capture_gate(
    const DoubleOKResult& result,
    const CaptureGateConfig& config,
    const std::optional<GlassesPose>& glasses_pose) {
    config.validate();
    const auto [pose_ok, pose_reason] = check_glasses_pose(glasses_pose, config);
    const int hand_count = static_cast<int>(result.hands.size());
    bool hands_visible = false;
    bool hands_centered = false;
    bool hands_separated = false;
    const bool double_ok = result.double_ok && (config.use_stable_double_ok ? result.stable_double_ok : true);

    if (!pose_ok) {
        return decision(pose_reason, config, pose_ok, hands_centered, hands_visible, hands_separated, double_ok, hand_count);
    }
    if (hand_count < 2) {
        return decision(
            GateReason::NeedTwoHands,
            config,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count);
    }

    const Landmarks& left = result.hands[0].landmarks;
    const Landmarks& right = result.hands[1].landmarks;
    validate_landmarks(left);
    validate_landmarks(right);

    hands_visible = is_hand_visible(left, config) && is_hand_visible(right, config);
    if (!hands_visible) {
        return decision(
            GateReason::HandsOutOfFrame,
            config,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count);
    }

    const Point2 left_center = center_of(left);
    const Point2 right_center = center_of(right);
    hands_centered = is_point_centered(left_center, config) && is_point_centered(right_center, config);
    if (!hands_centered) {
        return decision(
            GateReason::HandsNotCentered,
            config,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count);
    }

    hands_separated = distance(left_center, right_center) >= config.min_hand_separation;
    if (!hands_separated) {
        return decision(
            GateReason::HandsTooClose,
            config,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count);
    }

    if (config.require_double_ok && !double_ok) {
        return decision(
            GateReason::NeedDoubleOK,
            config,
            pose_ok,
            hands_centered,
            hands_visible,
            hands_separated,
            double_ok,
            hand_count);
    }

    return decision(
        GateReason::Ready,
        config,
        pose_ok,
        hands_centered,
        hands_visible,
        hands_separated,
        double_ok,
        hand_count);
}

CaptureGateDecision evaluate_labeled_capture_gate(
    const DoubleOKResult& result,
    const std::string& label,
    const CaptureGateConfig& config,
    const std::optional<GlassesPose>& glasses_pose) {
    if (label == "double_ok") {
        return evaluate_capture_gate(result, config, glasses_pose);
    }
    if (label != "not_double_ok") {
        throw std::invalid_argument("Unsupported capture label: " + label);
    }

    CaptureGateConfig geometry_config = config;
    geometry_config.require_double_ok = false;
    CaptureGateDecision geometry_decision = evaluate_capture_gate(result, geometry_config, glasses_pose);
    if (!geometry_decision.ready) {
        geometry_decision.gesture_ok = !result.double_ok;
        return geometry_decision;
    }
    if (result.double_ok) {
        return decision(
            GateReason::AvoidDoubleOK,
            geometry_config,
            geometry_decision.glasses_pose_ok,
            geometry_decision.hands_centered,
            geometry_decision.hands_visible,
            geometry_decision.hands_separated,
            result.double_ok,
            geometry_decision.hand_count,
            false);
    }
    return decision(
        GateReason::Ready,
        geometry_config,
        geometry_decision.glasses_pose_ok,
        geometry_decision.hands_centered,
        geometry_decision.hands_visible,
        geometry_decision.hands_separated,
        result.double_ok,
        geometry_decision.hand_count,
        true);
}

StereoCaptureGateDecision evaluate_stereo_capture_gate(
    const DoubleOKResult& left_result,
    const std::optional<DoubleOKResult>& right_result,
    const CaptureGateConfig& config,
    const std::optional<GlassesPose>& glasses_pose,
    StereoGateMode mode) {
    const CaptureGateDecision left_decision = evaluate_capture_gate(left_result, config, glasses_pose);
    if (mode == StereoGateMode::Left) {
        return {left_decision.ready, left_decision.reason, left_decision.prompt, left_decision, std::nullopt};
    }
    if (!right_result) {
        throw std::invalid_argument("right_result is required when stereo gate mode is both");
    }

    const CaptureGateDecision right_decision = evaluate_capture_gate(*right_result, config, glasses_pose);
    if (left_decision.ready && right_decision.ready) {
        return {true, GateReason::Ready, gate_prompt(GateReason::Ready), left_decision, right_decision};
    }
    if (!left_decision.ready && !right_decision.ready) {
        const std::string prompt = left_decision.reason == right_decision.reason
                                       ? "双目: " + left_decision.prompt
                                       : "左眼: " + left_decision.prompt + "; 右眼: " + right_decision.prompt;
        return {false, left_decision.reason, prompt, left_decision, right_decision};
    }
    const bool left_blocked = !left_decision.ready;
    const CaptureGateDecision blocked = left_blocked ? left_decision : right_decision;
    const std::string view = left_blocked ? "左眼" : "右眼";
    return {false, blocked.reason, view + ": " + blocked.prompt, left_decision, right_decision};
}

PromptSpeaker::PromptSpeaker(bool enabled, double min_interval_sec)
    : enabled_(enabled), min_interval_sec_(min_interval_sec) {
    if (!std::isfinite(min_interval_sec_) || min_interval_sec_ < 0.0) {
        throw std::invalid_argument("min_interval_sec must be finite and non-negative");
    }
}

void PromptSpeaker::emit(const std::string& prompt) {
    const double now = monotonic_seconds();
    if (prompt == last_prompt_ && now - last_time_ < min_interval_sec_) {
        return;
    }
    last_prompt_ = prompt;
    last_time_ = now;
    std::cout << prompt << '\n';
    if (enabled_) {
        std::cerr << "voice prompts are not spawned by the C++ build; prompt=" << prompt << '\n';
    }
}

}  // namespace double_ok_gesture
